File: sdks/python/endurance-tests/trends.py
"""Pure trend/count assertions for the endurance-tests suite: given the raw
numbers a run already collected, decide pass/fail. No FFI, no `reactor_sdk`,
no live service.

Split out of helpers.py (which does need all of that, to actually drive a
run) so these — the one place this suite's pass/fail *semantics* live — can
be unit-tested directly. helpers.py imports ../integration-tests/conftest.py
at module scope for its real-FFI plumbing (new_reactor, paced_connect, ...),
which makes `import helpers` require a built `reactor_sdk`; these functions
never needed that in the first place. See ../tests/test_trends.py.
"""

from __future__ import annotations

import dataclasses
import os
import statistics

from report import MetricResult

# Same knob as helpers.py's own copy of this line (that one re-exports this
# one) — see its docstring for why duration is wall-clock, not iteration
# count.
ENDURANCE_DURATION_SECONDS = float(os.environ.get("ENDURANCE_DURATION_SECONDS", "300"))


@dataclasses.dataclass
class Sample:
    cycle: int
    elapsed_s: float
    rss_bytes: int
    cpu_s: float
    # % of one CPU core busy since the *previous* sample (psutil's own
    # interval-based cpu_percent(), not an instantaneous reading) — a
    # different question from cpu_s (derived into cpu_deltas() below): that
    # one is "how much actual CPU work did this cycle do", this is "how busy
    # was the CPU relative to how long the cycle took". A cycle that's mostly
    # waiting on the network can do the same CPU work in more wall-clock time
    # and show a *lower* percentage even with identical cpu_s — the two can
    # disagree, and that disagreement itself is informative. Can exceed 100%
    # if more than one native thread is genuinely busy at once; that's
    # expected, not a bug, given this SDK's multi-threaded Rust runtime.
    cpu_percent: float
    live_clients: int
    orphaned_callbacks: int
    # OS-level, not SDK-level — a leaked native thread (the Rust runtime not
    # joining a worker) or a leaked socket/fd (a WebRTC connection not fully
    # torn down) is a classic class of bug in networking code that RSS alone
    # can miss for a while: a few thousand small, live allocations (thread
    # stacks, socket buffers) don't always show up as a dramatic RSS jump
    # before something else (an FD or thread ceiling) fails first.
    num_threads: int
    num_fds: int


def cpu_deltas(samples: list[Sample]) -> list[float]:
    """Per-interval CPU consumption, derived from `Sample.cpu_s`.

    `cpu_s` itself is `psutil`'s cumulative process CPU time since the
    process started — monotonically non-decreasing by construction, so a
    trend check on the raw values would always report growth regardless of
    whether anything is actually getting more expensive (the same mistake as
    using `resource.getrusage().ru_maxrss` for memory — see `ResourceSampler`'s
    own docstring). Diffing consecutive samples reframes it as CPU spent
    *per interval*, which is what can actually flag a cycle getting slower
    over the course of a run.
    """
    return [b.cpu_s - a.cpu_s for a, b in zip(samples, samples[1:])]


def assert_no_sustained_growth(
    values: list[float],
    *,
    name: str,
    max_growth_ratio: float,
    min_absolute_delta: float = 0.0,
    warmup_fraction: float = 0.2,
    use_median: bool = False,
    unit: str = "",
) -> MetricResult:
    """Fail if `values` is *still climbing in the back half of the run*: its
    last-third mean exceeds its middle-third mean (after dropping
    `warmup_fraction` to let one-time costs — import caches, connection
    setup, allocator warm-up — settle) by more than `max_growth_ratio`, *and*
    by more than `min_absolute_delta` in absolute terms.

    Deliberately last-vs-*middle*, not last-vs-first: a real CI run showed
    RSS flat, then a one-time ramp to a new plateau roughly in the middle of
    the window, then flat again for the rest of the run — a native
    buffer/pool growing once to its steady-state size, not an unbounded
    leak. Comparing against the *first* third means exactly where that one
    ramp happens to land is what decides pass/fail (whether the ramp's start
    lands inside the first-third window at all), which is timing, not
    signal — the same total jump measured as a comfortable pass or a
    razor-thin fail purely depending on when in the run it occurred, as
    happened across two otherwise-similar CI runs. Comparing the last third
    to the *middle* third instead asks the more direct question — "did it
    keep growing after that", not "is the end higher than the start" — so a
    one-time step that has already plateaued by the back third reads as flat
    (no fail), while something still genuinely climbing in the tail still
    trips it.

    The trade-off: a real leak that's still only in its early, slow-
    accelerating phase near the end of a short run could read as "already
    flat" and pass here where a first-vs-last comparison might have caught
    it. Preferred anyway — a slow leak has another cycle of this suite (or a
    longer ENDURANCE_DURATION_SECONDS run) to get caught once it's actually
    still climbing in *that* run's tail; a one-time step misread as a leak
    fails a real PR or blocks a release on nothing.

    `start`/`end`/`change` on the returned `MetricResult` (and the trend
    string in the printed verdict) still show the true first-third-to-last-
    third movement — that's still the right "how much did this move overall"
    number for a human reading the report — only the pass/fail decision
    itself is based on the middle-vs-last comparison.

    The absolute floor exists so a tiny, near-zero baseline can't turn an
    insignificant wobble into a ratio that looks huge. `values` must already
    be a *non-cumulative* measurement (raw RSS is fine as-is; CPU needs
    `cpu_deltas()` first — see its docstring). Not meant for exact-count
    fields (`live_clients`, `orphaned_callbacks`) — those are deterministic
    counts, not noisy measurements; use `assert_always_zero`/`assert_never_grows`
    for them instead.

    `use_median=True` swaps the mean for a median within each window —
    `num_threads` specifically has a documented one-cycle artifact (a cycle
    that catches the previous cycle's native thread teardown still in
    flight, briefly double-counting), and on a short run the "third" window
    can be small enough (as few as 2-3 samples) that a single such cycle
    landing in the last window skews its *mean* enough to misread as a
    sustained trend. A median shrugs off one outlier as long as it isn't the
    majority of the window; a real, sustained leak still moves the median
    just as surely as the mean.

    Always prints one line verdict, pass or fail — not just on failure — so a
    clean run still says *why* each signal looked fine, not just silence.

    Returns a `MetricResult` (status "ok"/"fail") rather than raising: the
    caller collects one from every check so a full report table can be built
    even when one of them fails, instead of stopping at the first failure.
    `n < 6` (not enough samples for a first/middle/last split at all) still
    raises directly — that's a test misconfiguration
    (ENDURANCE_DURATION_SECONDS too short), not a metric to report on.
    """
    n = len(values)
    if n < 6:
        raise AssertionError(
            f"only {n} {name} sample(s) collected in {ENDURANCE_DURATION_SECONDS}s — "
            "raise ENDURANCE_DURATION_SECONDS to get enough data for a trend"
        )
    warmed_up = values[int(n * warmup_fraction) :]
    third = max(1, len(warmed_up) // 3)
    first, middle, last = (
        warmed_up[:third],
        warmed_up[third : 2 * third],
        warmed_up[-third:],
    )
    average = statistics.median if use_median else (lambda xs: sum(xs) / len(xs))
    first_mean = average(first)
    middle_mean = average(middle)
    last_mean = average(last)

    # What decides pass/fail — see the docstring for why this is middle-vs-
    # last, not first-vs-last.
    tail_delta = last_mean - middle_mean
    tail_ratio = (tail_delta / middle_mean) if middle_mean else (1.0 if tail_delta > 0 else 0.0)
    is_leak = tail_delta > min_absolute_delta and tail_ratio > max_growth_ratio

    # What's reported to a human — the overall first-to-last movement, a
    # different (and both useful) number from what decided the verdict above.
    overall_delta = last_mean - first_mean
    overall_ratio = (
        (overall_delta / first_mean) if first_mean else (1.0 if overall_delta > 0 else 0.0)
    )

    trend = (
        f"{first_mean:,.3f} → {middle_mean:,.3f} → {last_mean:,.3f} "
        f"(overall {overall_ratio:+.0%}, tail {tail_ratio:+.0%})"
    )
    if is_leak:
        reason = (
            f"still climbing in the tail (last third {tail_ratio:+.0%} over the middle "
            f"third) — over the {max_growth_ratio:.0%} threshold, looks like a real leak"
        )
    elif tail_delta <= min_absolute_delta:
        reason = (
            f"the {tail_delta:,.3f} tail change is under the {min_absolute_delta:,.3g} floor, "
            f"so it's noise (or an already-settled one-time step) regardless of the "
            f"{tail_ratio:+.0%} tail ratio"
        )
    else:
        reason = f"tail growth is under the {max_growth_ratio:.0%} threshold — not still climbing"
    print(f"[{name}] {'LEAK?' if is_leak else 'ok'}: {trend} — {reason}")

    return MetricResult(
        name=name,
        start=first_mean,
        end=last_mean,
        change=overall_delta,
        unit=unit,
        status="fail" if is_leak else "ok",
        detail=reason,
        threshold=f"still climbing: last third > {max_growth_ratio:.0%} over middle third "
        f"(min {min_absolute_delta:,.3g})",
        mid=middle_mean,
    )


def assert_always_zero(samples: list[Sample], *, field: str, unit: str = "count") -> MetricResult:
    """Fail if `field` was ever nonzero, on *any* cycle — for a count that
    should return to exactly 0 every time (e.g. live client handles right
    after `close()`), a trend isn't the right test: a leak on cycle 3 that
    happens to get cleaned up by cycle 40 is still a real bug.

    Always prints one line verdict, pass or fail — see
    `assert_no_sustained_growth`'s own docstring for why it returns a
    `MetricResult` instead of raising directly.
    """
    bad = [(s.cycle, getattr(s, field)) for s in samples if getattr(s, field) != 0]
    last_value = getattr(samples[-1], field)
    if bad:
        cycle, value = bad[0]
        peak_value = max(v for _, v in bad)
        reason = (
            f"nonzero on {len(bad)}/{len(samples)} cycles (first at cycle {cycle}: "
            f"{value}, peak {peak_value}) — a handle leaked mid-run"
        )
        print(f"[{field}] LEAK?: {reason}")
        return MetricResult(
            name=field,
            start=0,
            # Report the peak, not the final sample: if `field` already
            # settled back to 0 by the last cycle, `last_value` would render
            # as "went from 0 to 0 (+0)" in the Markdown/text reports (which
            # don't include `detail`) and hide the mid-run leak entirely.
            end=peak_value,
            change=peak_value,
            unit=unit,
            status="fail",
            detail=reason,
            threshold="always 0",
            first_bad_cycle=cycle,
        )
    reason = f"stayed at exactly 0 across all {len(samples)} cycles"
    print(f"[{field}] ok: {reason}")
    return MetricResult(
        name=field,
        start=0,
        end=last_value,
        change=last_value,
        unit=unit,
        status="ok",
        detail=reason,
        threshold="always 0",
    )


def assert_never_grows(samples: list[Sample], *, field: str, unit: str = "count") -> MetricResult:
    """Fail if `field`'s peak ever exceeds its starting value — for a count
    that's expected to stay flat across the whole run (not necessarily 0).

    Always prints one line verdict, pass or fail — see
    `assert_no_sustained_growth`'s own docstring for why it returns a
    `MetricResult` instead of raising directly.
    """
    baseline = getattr(samples[0], field)
    peak = max(getattr(s, field) for s in samples)
    if peak > baseline:
        reason = f"grew from {baseline} to {peak} during the run"
        print(f"[{field}] LEAK?: {reason}")
        return MetricResult(
            name=field,
            start=baseline,
            end=peak,
            change=peak - baseline,
            unit=unit,
            status="fail",
            detail=reason,
            threshold=f"never above {baseline}",
        )
    reason = f"never exceeded its starting value ({baseline}; peak seen was {peak})"
    print(f"[{field}] ok: {reason}")
    return MetricResult(
        name=field,
        start=baseline,
        end=peak,
        change=peak - baseline,
        unit=unit,
        status="ok",
        detail=reason,
        threshold=f"never above {baseline}",
    )


def standard_resource_metrics(
    samples: list[Sample],
    *,
    live_clients_baseline_zero: bool = False,
    fds_exact: bool = False,
) -> list[MetricResult]:
    """The RSS/CPU/thread/fd/handle checks every scenario in this suite runs
    at the end of its loop — live_clients, orphaned_callbacks, rss,
    cpu_s_per_cycle, cpu_percent, num_threads, num_fds — with the same
    thresholds every scenario has already converged on independently.
    Extracted so a new scenario is "write the loop body, call this," not
    another copy of this same ~40-line block: see ../README.md's "Adding a
    new scenario" and the sdk-from-ffi skill's "Endurance and leak tests"
    section for the shape this is meant to support.

    `live_clients_baseline_zero`: True for a scenario that starts with no
    connected client (a fresh `Reactor` per cycle, e.g. lifecycle-churn) —
    `live_clients` must be **exactly** 0 after every cycle
    (`assert_always_zero`). False (the default) for a scenario built around
    one long-lived session connected once by the `reactor` fixture — its
    baseline is 1, and `live_clients` must never exceed that
    (`assert_never_grows`).

    `fds_exact`: True when fd teardown is synchronous with this scenario's
    own cycle boundary (only lifecycle-churn's disconnect()/close() per
    cycle proved this in practice) — `num_fds` must never exceed its
    starting value (`assert_never_grows`). False (the default) for anything
    that can legitimately open a few more during warm-up and then plateau
    (a connection pool, a worker thread) — `num_fds` gets the same
    trend-based check as RSS/CPU instead.

    A scenario with its own extra invariants (session-churn's
    `pending_completions`/`Track._adapters`, publish-churn's `track.published`,
    pause-resume-churn's `reactor.paused_tracks`) still checks those itself,
    in its own loop — they are specific to what that scenario exercises, not
    generic resource accounting, so they do not belong here.
    """
    metrics: list[MetricResult] = []
    if live_clients_baseline_zero:
        metrics.append(assert_always_zero(samples, field="live_clients"))
    else:
        metrics.append(assert_never_grows(samples, field="live_clients"))
    metrics.append(assert_always_zero(samples, field="orphaned_callbacks"))
    if fds_exact:
        metrics.append(assert_never_grows(samples, field="num_fds"))
    metrics.append(
        assert_no_sustained_growth(
            [s.rss_bytes / 1e6 for s in samples],
            name="rss",
            unit="MB",
            max_growth_ratio=0.15,
            min_absolute_delta=5.0,
        )
    )
    metrics.append(
        assert_no_sustained_growth(
            cpu_deltas(samples),
            name="cpu_s_per_cycle",
            unit="s",
            max_growth_ratio=0.5,
            min_absolute_delta=0.05,
        )
    )
    metrics.append(
        assert_no_sustained_growth(
            [s.cpu_percent for s in samples],
            name="cpu_percent",
            unit="%",
            max_growth_ratio=0.5,
            min_absolute_delta=5.0,
        )
    )
    metrics.append(
        assert_no_sustained_growth(
            [float(s.num_threads) for s in samples],
            name="num_threads",
            unit="count",
            max_growth_ratio=0.15,
            min_absolute_delta=4,
            use_median=True,
        )
    )
    if not fds_exact:
        metrics.append(
            assert_no_sustained_growth(
                [float(s.num_fds) for s in samples],
                name="num_fds",
                unit="count",
                max_growth_ratio=0.15,
                min_absolute_delta=3,
            )
        )
    return metrics

File: sdks/cpp/endurance-tests/trends.cpp
#include "trends.hpp"

#include <algorithm>
#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <numeric>
#include <sstream>
#include <stdexcept>

namespace endurance {

double endurance_duration_seconds() {
  const char* raw = std::getenv("ENDURANCE_DURATION_SECONDS");
  return raw != nullptr ? std::stod(raw) : 300.0;
}

std::vector<double> cpu_deltas(const std::vector<Sample>& samples) {
  std::vector<double> deltas;
  if (samples.size() < 2) {
    return deltas;
  }
  deltas.reserve(samples.size() - 1);
  for (std::size_t i = 1; i < samples.size(); ++i) {
    deltas.push_back(samples[i].cpu_s - samples[i - 1].cpu_s);
  }
  return deltas;
}

namespace {

double mean(const std::vector<double>& xs) {
  return std::accumulate(xs.begin(), xs.end(), 0.0) / static_cast<double>(xs.size());
}

double median(std::vector<double> xs) {
  std::sort(xs.begin(), xs.end());
  const std::size_t mid = xs.size() / 2;
  return xs.size() % 2 == 0 ? (xs[mid - 1] + xs[mid]) / 2.0 : xs[mid];
}

std::string format_ratio(double ratio) {
  std::ostringstream oss;
  oss << (ratio >= 0 ? "+" : "") << std::fixed << std::setprecision(0) << (ratio * 100) << "%";
  return oss.str();
}

}  // namespace

MetricResult assert_no_sustained_growth(const std::vector<double>& values, const std::string& name,
                                        double max_growth_ratio, double min_absolute_delta,
                                        double warmup_fraction, bool use_median,
                                        const std::string& unit) {
  const std::size_t n = values.size();
  if (n < 6) {
    throw std::runtime_error("only " + std::to_string(n) + " " + name +
                             " sample(s) collected — raise ENDURANCE_DURATION_SECONDS to get "
                             "enough data for a trend");
  }
  const std::vector<double> warmed_up(
      values.begin() + static_cast<std::ptrdiff_t>(static_cast<double>(n) * warmup_fraction),
      values.end());
  const std::size_t third = std::max<std::size_t>(1, warmed_up.size() / 3);
  const std::vector<double> first(warmed_up.begin(),
                                  warmed_up.begin() + static_cast<std::ptrdiff_t>(third));
  const std::vector<double> middle(warmed_up.begin() + static_cast<std::ptrdiff_t>(third),
                                   warmed_up.begin() + static_cast<std::ptrdiff_t>(2 * third));
  const std::vector<double> last(warmed_up.end() - static_cast<std::ptrdiff_t>(third),
                                 warmed_up.end());

  const auto average = [use_median](const std::vector<double>& xs) {
    return use_median ? median(xs) : mean(xs);
  };
  const double first_mean = average(first);
  const double middle_mean = average(middle);
  const double last_mean = average(last);

  // What decides pass/fail — see this function's own doc-comment (trends.hpp)
  // for why this is middle-vs-last, not first-vs-last.
  const double tail_delta = last_mean - middle_mean;
  const double tail_ratio =
      middle_mean != 0.0 ? (tail_delta / middle_mean) : (tail_delta > 0 ? 1.0 : 0.0);
  const bool is_leak = tail_delta > min_absolute_delta && tail_ratio > max_growth_ratio;

  // What's reported to a human — the overall first-to-last movement, a
  // different (and both useful) number from what decided the verdict above.
  const double overall_delta = last_mean - first_mean;
  const double overall_ratio =
      first_mean != 0.0 ? (overall_delta / first_mean) : (overall_delta > 0 ? 1.0 : 0.0);

  std::ostringstream trend_stream;
  trend_stream << std::fixed << std::setprecision(3) << first_mean << " → " << middle_mean << " → "
               << last_mean << " (overall " << format_ratio(overall_ratio) << ", tail "
               << format_ratio(tail_ratio) << ")";
  const std::string trend = trend_stream.str();

  std::string reason;
  if (is_leak) {
    reason = "still climbing in the tail (last third " + format_ratio(tail_ratio) +
             " over the middle third) — over the " +
             std::to_string(static_cast<int>(max_growth_ratio * 100)) +
             "% threshold, looks like a real leak";
  } else if (tail_delta <= min_absolute_delta) {
    std::ostringstream floor_stream;
    floor_stream << std::setprecision(3) << tail_delta;
    reason = "the " + floor_stream.str() + " tail change is under the " +
             std::to_string(min_absolute_delta) +
             " floor, so it's noise (or an already-settled one-time step) regardless of the " +
             format_ratio(tail_ratio) + " tail ratio";
  } else {
    reason = "tail growth is under the " +
             std::to_string(static_cast<int>(max_growth_ratio * 100)) +
             "% threshold — not still climbing";
  }
  std::cout << "[" << name << "] " << (is_leak ? "LEAK?" : "ok") << ": " << trend << " — " << reason
            << "\n";

  MetricResult result;
  result.name = name;
  result.start = first_mean;
  result.end = last_mean;
  result.change = overall_delta;
  result.unit = unit;
  result.status = is_leak ? "fail" : "ok";
  result.detail = reason;
  result.threshold = "still climbing: last third > " +
                     std::to_string(static_cast<int>(max_growth_ratio * 100)) +
                     "% over middle third";
  result.mid = middle_mean;
  return result;
}

MetricResult assert_always_zero(const std::vector<Sample>& samples,
                                const std::function<long(const Sample&)>& get,
                                const std::string& name, const std::string& unit) {
  std::vector<std::pair<int, long>> bad;
  for (const auto& s : samples) {
    const long value = get(s);
    if (value != 0) {
      bad.emplace_back(s.cycle, value);
    }
  }
  const long last_value = get(samples.back());

  MetricResult result;
  result.name = name;
  result.unit = unit;
  result.start = 0;
  result.threshold = "always 0";

  if (!bad.empty()) {
    const auto [cycle, value] = bad.front();
    long peak_value = value;
    for (const auto& [c, v] : bad) {
      peak_value = std::max(peak_value, v);
    }
    const std::string reason = "nonzero on " + std::to_string(bad.size()) + "/" +
                               std::to_string(samples.size()) + " cycles (first at cycle " +
                               std::to_string(cycle) + ": " + std::to_string(value) + ", peak " +
                               std::to_string(peak_value) + ") — a handle leaked mid-run";
    std::cout << "[" << name << "] LEAK?: " << reason << "\n";
    result.end = static_cast<double>(peak_value);
    result.change = static_cast<double>(peak_value);
    result.status = "fail";
    result.detail = reason;
    result.first_bad_cycle = cycle;
    return result;
  }

  const std::string reason =
      "stayed at exactly 0 across all " + std::to_string(samples.size()) + " cycles";
  std::cout << "[" << name << "] ok: " << reason << "\n";
  result.end = static_cast<double>(last_value);
  result.change = static_cast<double>(last_value);
  result.status = "ok";
  result.detail = reason;
  return result;
}

MetricResult assert_never_grows(const std::vector<Sample>& samples,
                                const std::function<long(const Sample&)>& get,
                                const std::string& name, const std::string& unit) {
  const long baseline = get(samples.front());
  long peak = baseline;
  for (const auto& s : samples) {
    peak = std::max(peak, get(s));
  }

  MetricResult result;
  result.name = name;
  result.unit = unit;
  result.start = static_cast<double>(baseline);
  result.end = static_cast<double>(peak);
  result.change = static_cast<double>(peak - baseline);
  result.threshold = "never above " + std::to_string(baseline);

  if (peak > baseline) {
    const std::string reason =
        "grew from " + std::to_string(baseline) + " to " + std::to_string(peak) + " during the run";
    std::cout << "[" << name << "] LEAK?: " << reason << "\n";
    result.status = "fail";
    result.detail = reason;
    return result;
  }
  const std::string reason = "never exceeded its starting value (" + std::to_string(baseline) +
                             "; peak seen was " + std::to_string(peak) + ")";
  std::cout << "[" << name << "] ok: " << reason << "\n";
  result.status = "ok";
  result.detail = reason;
  return result;
}

std::vector<MetricResult> standard_resource_metrics(const std::vector<Sample>& samples) {
  std::vector<MetricResult> metrics;

  metrics.push_back(assert_no_sustained_growth(
      [&] {
        std::vector<double> rss;
        rss.reserve(samples.size());
        for (const auto& s : samples) {
          rss.push_back(static_cast<double>(s.rss_bytes) / 1e6);
        }
        return rss;
      }(),
      "rss", 0.15, 5.0, 0.2, /*use_median=*/false, "MB"));

  metrics.push_back(assert_no_sustained_growth(cpu_deltas(samples), "cpu_s_per_cycle", 0.5, 0.05,
                                               0.2, /*use_median=*/false, "s"));

  metrics.push_back(assert_no_sustained_growth(
      [&] {
        std::vector<double> pct;
        pct.reserve(samples.size());
        for (const auto& s : samples) {
          pct.push_back(s.cpu_percent);
        }
        return pct;
      }(),
      "cpu_percent", 0.5, 5.0, 0.2, /*use_median=*/false, "%"));

  // Median-backed, both of these — see this function's own doc-comment
  // (trends.hpp) for why num_fds joins num_threads here rather than getting
  // Python's exact "never past its starting value" check.
  metrics.push_back(assert_no_sustained_growth(
      [&] {
        std::vector<double> threads;
        threads.reserve(samples.size());
        for (const auto& s : samples) {
          threads.push_back(static_cast<double>(s.num_threads));
        }
        return threads;
      }(),
      "num_threads", 0.15, 4.0, 0.2, /*use_median=*/true, "count"));

  metrics.push_back(assert_no_sustained_growth(
      [&] {
        std::vector<double> fds;
        fds.reserve(samples.size());
        for (const auto& s : samples) {
          fds.push_back(static_cast<double>(s.num_fds));
        }
        return fds;
      }(),
      "num_fds", 0.15, 3.0, 0.2, /*use_median=*/true, "count"));

  return metrics;
}

}  // namespace endurance
